statistics/cmsketch.go

```diff
@@ -44,29 +60,173 @@ func NewCMSketch(d, w int32) *CMSketch {
+	rowCount = mathutil.MaxUint64(rowCount, uint64(len(sample)))
+	estimateNDV, scaleRatio := calculateEstimateNDV(helper, rowCount)
+	c := buildCMSWithTopN(helper, d, w, scaleRatio)
+	c.calculateDefaultVal(helper, estimateNDV, scaleRatio, rowCount)
+	return c
+}
+
+func buildCMSWithTopN(helper *topNHelper, d, w int32, scaleRatio uint64) (c *CMSketch) {
+	c, helper.sumTopN, helper.numTop = NewCMSketch(d, w), 0, 0
+	enableTopN := helper.sampleSize/topNThreshold <= helper.sumTopN
+	if enableTopN {
+		c.topN = make(map[uint64][]topNMeta)
+	}
+	for counterKey, cnt := range helper.counter {
+		data, scaledCount := hack.Slice(string(counterKey)), cnt*scaleRatio
+		if enableTopN && cnt >= helper.lastVal {
+			h1, h2 := murmur3.Sum128(data)
+			c.topN[h1] = append(c.topN[h1], topNMeta{h1, h2, data, scaledCount})
+			helper.sumTopN += scaledCount
+			helper.numTop++
+		} else {
+			c.updateBytesWithDelta(data, scaledCount)
+		}
+	}
+	return
+}
+
+func (c *CMSketch) calculateDefaultVal(helper *topNHelper, estimateNDV, scaleRatio, rowCount uint64) {
+	sampleNDV := uint64(len(helper.sorted))
+	if rowCount <= helper.sumTopN {
+		c.defaultValue = 1
+	} else if estimateNDV <= uint64(helper.numTop) {
+		c.defaultValue = 1
+	} else if estimateNDV+helper.onlyOnceItems <= uint64(sampleNDV) {
+		c.defaultValue = 1
+	} else {
+		estimateRemainingCount := rowCount - (helper.sampleSize-uint64(helper.onlyOnceItems))*scaleRatio
+		c.defaultValue = estimateRemainingCount / (estimateNDV - uint64(sampleNDV) + helper.onlyOnceItems)
+	}
+}
+
+// queryAddTopN TopN adds count to CMSketch.topN if exists, and returns the count of such elements after insert.
+// If such elements does not in topn elements, nothing will happen and false will be returned.
+func (c *CMSketch) updateTopNWithDelta(h1, h2 uint64, d []byte, delta uint64) bool {
+	if c.topN == nil {
+		return false
+	}
+	for _, cnt := range c.topN[h1] {
+		if cnt.h2 == h2 && bytes.Equal(d, cnt.data) {
+			cnt.count += delta
+			return true
+		}
+	}
+	return false
+}
+
+func (c *CMSketch) queryTopN(h1, h2 uint64, d []byte) (uint64, bool) {
+	if c.topN == nil {
+		return 0, false
+	}
+	for _, cnt := range c.topN[h1] {
+		if cnt.h2 == h2 && bytes.Equal(d, cnt.data) {
+			return cnt.count, true
+		}
+	}
+	return 0, false
+}
+
 // InsertBytes inserts the bytes value into the CM Sketch.
 func (c *CMSketch) InsertBytes(bytes []byte) {
-	c.count++
+	c.updateBytesWithDelta(bytes, 1)
+}
+
+// updateBytesWithDelta adds the bytes value into the CM Sketch by delta.
+func (c *CMSketch) updateBytesWithDelta(bytes []byte, delta uint64) {
 	h1, h2 := murmur3.Sum128(bytes)
+	if c.updateTopNWithDelta(h1, h2, bytes, delta) {
+		return
+	}
+	c.count += delta
 	for i := range c.table {
 		j := (h1 + h2*uint64(i)) % uint64(c.width)
-		c.table[i][j]++
+		c.table[i][j] += uint32(delta)
 	}
 }
 
+func (c *CMSketch) considerDefVal(cnt uint64) bool {
+	return cnt < 2*(c.count/uint64(c.width)) && c.defaultValue > 0
```

Comment:
What is the rationale behind this formula?

Reply:
Like line 154, scaled count might be too big for some element, we should consider return default value.